File: 0021-merge-two-sorted-lists/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if (l1 == NULL)
            return l2;
        if (l2 == NULL)
            return l1;
        
        if (l1->val > l2->val)
        {
            // Ensure l1 is smaller one.
            ListNode* tmp_node = l1;
            l1 = l2;
            l2 = tmp_node;
        }
        ListNode* cur_node1 = l1;
        ListNode* cur_node2 = l2;
        
        // Loop until all two pointers reach the end.
        while(cur_node1 != NULL)
        {
            if (cur_node1->next == NULL)
            {
                cur_node1->next = cur_node2;
                break;    // break if the end of list 1 is reached.
            }
            
            if (cur_node2 == NULL)
                break;    // break if the end of list 2 is reached.  
            
            // Append l2 to l1.
            ListNode* next_node1 = cur_node1->next;
            if ((cur_node1->val <= cur_node2->val) && (next_node1->val > cur_node2->val))
            {
                cur_node1->next = cur_node2;
                
                ListNode* next_node2 = cur_node2->next;
                while ((next_node2 != NULL) && (next_node2->val < next_node1->val))
                {
                    cur_node2 = cur_node2->next;
                    next_node2 = cur_node2->next;
                }
                std::cout << cur_node2->val << next_node1->val << std::endl;
                    
                cur_node2->next = next_node1;
                
                cur_node2 = next_node2;
            }
            cur_node1 = next_node1;
            
        }
        return l1;
    }
};

File: 0021-merge-two-sorted-lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        if (not l1) or (not l2):
            return l1 or l2
        new_l = None
        cur_l = None
        l1_cur = l1
        l2_cur = l2
        while l1_cur and l2_cur:
            if l1_cur.val <= l2_cur.val:
                tmp_l1_cur = l1_cur.next
                l1_cur.next = None
                cur_l_next = l1_cur
                l1_cur = tmp_l1_cur
            else:
                tmp_l2_cur = l2_cur.next
                l2_cur.next = None
                cur_l_next = l2_cur
                l2_cur = tmp_l2_cur
            if new_l is None:
                new_l = cur_l_next
            if cur_l is None:
                cur_l = cur_l_next
            else:
                cur_l.next = cur_l_next
                cur_l = cur_l.next
        if l1_cur:
            cur_l.next = l1_cur
        if l2_cur:
            cur_l.next = l2_cur
        return new_l
